corehq/apps/reports/view_helpers.py

```diff
@@ -0,0 +1,231 @@
+            'form_url_root': form_url_root,
+            'view_url': get_case_url(case.case_id),
+            'session_data': get_session_data(case, current_case, type_info)
+        }
+        data = type_info.get(case.type, {})
+        if 'description_property' in data:
+            data['description'] = getattr(case, data['description_property'], None)
+        if 'edit_session_data' in data:
+            data['session_data'].update(data['edit_session_data'])
+        data.update(common_data)
+
+        case.edit_data = data
+
+        if 'child_type' in data and not case.closed:
+            child_type = data['child_type']
+            child_data = type_info.get(child_type, {})
+            child_data.update(common_data)
+            child_data.update({
+                "link_text": _("Add %(case_type)s") % {
+                    'case_type': child_data.get('type_name', child_type)
+                },
+                "parent_node_id": case.case_id,
+            })
+
+            if 'create_session_data' in child_data:
+                child_data['session_data'].update(child_data['create_session_data'])
+            case.add_child_data = child_data
+
+    process_output(case_output)
+
+
+def get_case_hierarchy(case, type_info):
+    def get_children(case, referenced_type=None, seen=None):
+        seen = seen or set()
+
+        ignore_types = type_info.get(case.type, {}).get("ignore_relationship_types", [])
+        if referenced_type and referenced_type in ignore_types:
+            return None
+
+        seen.add(case.case_id)
+        children = [
+            get_children(i.referenced_case, i.referenced_type, seen) for i in case.reverse_indices
+            if i.referenced_id not in seen
+        ]
+
+        children = [c for c in children if c is not None]
+
+        # non-first-level descendants
+        descendant_types = []
+        for c in children:
+            descendant_types.extend(c['descendant_types'])
+        descendant_types = list(set(descendant_types))
+
+        children = sorted(children, key=partial(sortkey, type_info=type_info))
+
+        # set parent_case_id used by flat display
+        for c in children:
+            if not hasattr(c['case'], 'treetable_parent_node_id'):
+                c['case'].treetable_parent_node_id = case.case_id
+
+        child_cases = []
+        for c in children:
+            child_cases.extend(c['case_list'])
+
+        return {
+            'case': case,
+            'child_cases': children,
+            'descendant_types': list(set(descendant_types + [c['case'].type for c in children])),
+            'case_list': [case] + child_cases
+        }
+
+    return get_children(case)
+
+
+def get_flat_descendant_case_list(case, get_case_url, type_info=None):
+    type_info = type_info or {}
+    hierarchy = get_case_hierarchy(case, type_info)
+    process_case_hierarchy(hierarchy, get_case_url, type_info)
+    return hierarchy['case_list']
```

Comment:
I didn't really review this file as it looks like it is just straight copy/paste. Let me know if I should

Reply:
nope, it's indeed copy/paste